mesh/submesh/submesh.hpp

```diff
@@ -0,0 +1,201 @@
+   {
+      Domain,
+      Boundary
+   };
+
+   static const int GENERATED_ATTRIBUTE = 900;
+
+   SubMesh() = delete;
+
+   /**
+    * @brief Create a domain SubMesh from it's parent.
+    *
+    * The SubMesh object expects the parent Mesh object to be valid for the
+    * entire object lifetime. The @a domain_attributes have to mark exactly one
+    * connected subset of the parent Mesh.
+    *
+    * @param[in] parent Parent Mesh
+    * @param[in] domain_attributes Domain attributes to extract
+    */
+   static SubMesh CreateFromDomain(const Mesh &parent,
+                                   Array<int> domain_attributes);
+
+   /**
+   * @brief Create a surface SubMesh from it's parent.
+   *
+   * The SubMesh object expects the parent Mesh object to be valid for the
+   * entire object lifetime. The @a boundary_attributes have to mark exactly one
+   * connected subset of the parent Mesh.
+   *
+   * @param[in] parent Parent Mesh
+   * @param[in] boundary_attributes Boundary attributes to extract
+
+   */
+   static SubMesh CreateFromBoundary(const Mesh &parent,
+                                     Array<int> boundary_attributes);
+
+   /**
+    * @brief Get the parent Mesh object
+    *
+    */
+   const Mesh* GetParent() const
+   {
+      return &parent_;
+   }
+
+   /**
+    * @brief Get the From indicator.
+    *
+    * Indicates whether the SubMesh has been created from a domain or
+    * surface.
+    */
+   const From GetFrom() const
+   {
+      return from_;
+   }
+
+   /**
+    * @brief Get the parent element id map.
+    *
+    * SubMesh element id (array index) to parent Mesh element id.
+    */
+   const Array<int>& GetParentElementIDMap() const
+   {
+      return parent_element_ids_;
+   }
+
+   /**
+    * @brief Get the face id map
+    *
+    * SubMesh element id (array index) to parent Mesh face id.
+    */
+   const Array<int>& GetParentFaceIDMap() const
+   {
+      return parent_face_ids_;
+   }
+
+   /**
+    * @brief Get the parent vertex id map.
+    *
+    * SubMesh vertex id (array index) to parent Mesh vertex id.
+    */
+   const Array<int>& GetParentVertexIDMap() const
+   {
+      return parent_vertex_ids_;
+   }
+
+   /**
+    * @brief Transfer the dofs of a GridFunction.
+    *
+    * The @a src GridFunction can either be defined on a Mesh or a SubMesh and
+    * is transferred appropriately.
+    *
+    * @note Either @a src or @a dst has to be defined on a SubMesh.
+    *
+    * @param[in] src
+    * @param[out] dst
+    */
+   static void Transfer(const GridFunction &src, GridFunction &dst);
+
+   /**
+    * @brief Create a Transfer Map object.
+    *
+    * The @a src GridFunction can either be defined on a Mesh or a
+    * SubMesh and is transferred appropriately.
+    *
+    * The returned pointer can be stored, reused and must be deleted.
+    *
+    * @note Either @a src or @a dst has to be defined on a SubMesh.
+    */
+   static const TransferMap* CreateTransferMap(const GridFunction &src,
```

Comment:
Can you return the object instead of a pointer using guaranteed copy elision? I forget what version of the standard that came through in....

Reply:
C++17 :neutral_face:

Reply:
It seems to work with clang and c++11. So I'm going to change the behavior here.